src/consumer/runner.js

```diff
@@ -297,22 +307,62 @@ module.exports = class Runner {
     }
 
     const concurrently = limitConcurrency({ limit: this.partitionsConsumedConcurrently })
-    await Promise.all(
-      batches.map(batch =>
-        concurrently(async () => {
-          if (!this.running) {
-            return
-          }
-
-          if (batch.isEmpty()) {
-            return
-          }
-
-          await onBatch(batch)
-        })
-      )
-    )
 
+    if (!this.running) {
+      return
+    }
+
+    const enqueuedTasks = []
+    let expectedNumberOfExecutions = 0
+    let numberOfExecutions = 0
+    const { lock, unlock, unlockWithError } = barrier()
+
+    while (true) {
+      const result = iterator.next()
+      if (result.done) {
+        break
+      }
+
+      enqueuedTasks.push(async () => {
+        if (!this.running) {
+          return
+        }
+
+        const batches = await result.value
+        expectedNumberOfExecutions += batches.length
+
+        batches.map(batch =>
+          concurrently(async () => {
+            try {
+              if (!this.running) {
+                return
+              }
+
+              if (batch.isEmpty()) {
+                return
+              }
+
+              await onBatch(batch)
+              await this.consumerGroup.heartbeat({ interval: this.heartbeatInterval })
```

Comment:
@tulios calling heartbeat here, will fix this issue: https://github.com/tulios/kafkajs/issues/708

But, in case of `eachMessage`, we already call `heartbeat()` [inside processEachMessage()](https://github.com/tulios/kafkajs/blob/master/src/consumer/runner.js#L181) method.
With this, we'll do two heartbeats after every message.

Reply:
Okay, after digging in more code, it looks this call will not effect `eachMessage()` because of `heartbeatInterval`. Still, it'd be good to keep it only at one place.
Maybe I don't have full context here.